### PR TITLE
Repo reformatting

### PR DESCRIPTION
**Pull request by dkadochnikov**
Change of the repository structure for better code understanding after the implementation of future tasks. Pipeline only works for classification task for now.

### CODE REVIEW

File: src/models/backbone.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    hidden_size: int = 768,
+    num_attention_heads: int = 12,
+    num_hidden_layers: int = 1,
+    type_vocab_size: int = 1,
+    dropout: float = 0.0,
+    num_classes: int = 2,
+):
+    if vtr:
+        logger.info(f"Initializing VTR classifier | hidden size: {hidden_size}, # layers: {num_hidden_layers}")
+        return nn.Linear(hidden_size, num_classes)
```
Comment (SpirinEgor) [inline src/models/backbone.py:19]: This is strange, regardless embedder, backbone should be the same: VTR + BertForSequenceClassification or TTR + BertForSequenceClassification

File: src/models/embedders/ttr.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from torch import nn
+
+
+class VanillaEmbedder(nn.Module):
```
Comment (SpirinEgor) [inline src/models/embedders/ttr.py:4]: `TTREmbedder`

File: src/models/embedders/ttr.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, batch):
+        output = {"embeddings": [batch["input_ids"], batch["attention_mask"], batch["token_type_ids"]]}
```
Comment (SpirinEgor) [inline src/models/embedders/ttr.py:9]: No, this is not what we want. You should apply `nn.Embeddings` layer to `batch["input_ids"]` here and return `[batch size; seq len; emb dim]`

File: src/models/embedders/vtr.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    stride: int = 1,
+    dilation: int = 1,
+):
+    return nn.Sequential(
+        ResBlock(in_channels, out_channels, conv_kernel_size, padding, stride, dilation),
+        nn.MaxPool2d(pool_kernel_size),
+    )
+
+
+class VisualEmbedder(nn.Module):
```
Comment (SpirinEgor) [inline src/models/embedders/vtr.py:93]: `VTREmbedder`

File: src/models/embedders/vtr.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+        encoder_layer = nn.TransformerEncoderLayer(
+            d_model=hidden_size,
+            dim_feedforward=hidden_size * 4,
+            nhead=num_attention_heads,
+            dropout=dropout,
+            batch_first=True,
+        )
+        self.encoder = nn.TransformerEncoder(encoder_layer, num_layers=num_layers)
+        self.norm = nn.LayerNorm(hidden_size)
```
Comment (SpirinEgor) [inline src/models/embedders/vtr.py:137]: This is not the part of the embedder, this is complete model

File: src/models/tasks.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+from src.models.vtr.ocr import OCRHead
+
+
+class ToxicClassifier(nn.Module):
+    def __init__(
+        self,
+        classifier,
+        embedder,
+        vtr: bool = False,
```
Comment (SpirinEgor) [inline src/models/tasks.py:13]: I guess this is unnecessary for this class

File: src/models/tasks.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+from typing import Type
+import torch
+from torch import nn
+
+from src.models.vtr.ocr import OCRHead
+
+
+class ToxicClassifier(nn.Module):
```
Comment (SpirinEgor) [inline src/models/tasks.py:8]: `SequenceClassifier`

**Review by SpirinEgor — COMMENTED** [review]
Well, nice start, but we should properly discuss structure and model decomposition

File: src/main.py [hunk 8]
```diff
@@ -55,14 +57,21 @@ def train_vanilla_encoder(args: Namespace, train_data: list, val_data: list = No
...
 
-    model = TokensToxicClassifier(
-        vocab_size=train_dataset.tokenizer.vocab_size,
-        num_layers=model_config.num_layers,
-        hidden_size=model_config.emb_size,
-        num_classes=model_config.num_classes,
-        num_attention_heads=model_config.n_head,
-        dropout=model_config.dropout,
-    )
+    backbone_config = {
```
Comment (SpirinEgor) [inline src/main.py:60]: You should use model_config directly, no need in additional wrapping

File: src/main.py [hunk 9]
```diff
@@ -55,14 +57,21 @@ def train_vanilla_encoder(args: Namespace, train_data: list, val_data: list = No
...
+        "vocab_size": train_dataset.tokenizer.vocab_size,
+        "hidden_size": model_config.emb_size,
+        "num_hidden_layers": model_config.num_layers,
+        "num_attention_heads": model_config.n_head,
+        "hidden_dropout_prob": model_config.dropout,
+        "attention_probs_dropout_prob": model_config.dropout,
+        "max_position_embeddings": training_config.max_seq_len,
+        "num_classes": model_config.num_classes,
+        # type_vocab_size - ?
```
Comment (SpirinEgor) [inline src/main.py:69]: We don't have any types, no need to specify them

File: src/models/embedders/vtr.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        conv = self.slice_conv(batch["slices"].view(batch_size * slice_count, 1, height, width))
+
+        _, channels_count, h_out, w_out = conv.shape
+        batched_conv = conv.view(batch_size, slice_count, channels_count * h_out * w_out)
+        batched_conv = self.linear_bridge(batched_conv)
+
+        output = {"embeddings": batched_conv, "ocr_embeddings": conv}
+
+        return output  # [batch size, slice count, emb size],
+        # [batch size * slice count, out channels, emb height, emb width]
```
Comment (SpirinEgor) [inline src/models/embedders/vtr.py:131]: Adding info about tensor shapes is very useful, but let's do it in obvious way 😅
```python
return {
    "embeddings": batched_conv,  # [batch size, slice count, emb size]
    "convolutions": conv,   # [batch size * slice count, out channels, emb height, emb width]
}
```

Comment (SpirinEgor) [reply src/models/embedders/vtr.py:131]: Btw, maybe it is more native to return convolutions with shape: `[batch size; slice count; out_channels; h; w]`?

File: src/models/tasks.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+        logger.info(
+            f"Initializing vanilla BERT classifier | hidden size: {config['hidden_size']}, "
+            f"# layers: {config['num_hidden_layers']}"
+        )
+
+        model_config = BertConfig(**config)
+        self.backbone = BertForSequenceClassification(model_config)
+        self.embedder = embedder
+        self.ocr = ocr
```
Comment (SpirinEgor) [inline src/models/tasks.py:22]: Please, add here an additional check: if `ocr` is passed then the embedder should be an instance of `VTREmbedder`

File: src/models/tasks.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        result = {}
+        output = self.embedder(input_batch)  # batch_size, seq_len, emb_size
+        output["embeddings"] = self.positional(output["embeddings"])
+        if self.ocr:
+            result["ocr_logits"] = self.ocr(output["ocr_embeddings"])
+
+        logits = self.backbone(inputs_embeds=output["embeddings"])  # batch_size, num_classes
+        result["logits"] = logits
+
+        return result
```
Comment (SpirinEgor) [inline src/models/tasks.py:37]: Let's return full output from the backbone, it will also contain loss if you pass labels.
```python
embeddings = self.embedder(input_batch)  # batch_size, seq_len, emb_size

model_input = self.positional(output["embeddings"])
result = self.backbone(inputs_embeds=model_input)

if self.ocr is not None:
    result["ocr_logits] = self.ocr(output["convolutions"])
```

File: src/models/tasks.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+from torch.nn.utils.rnn import pad_sequence
+from torch.nn import CTCLoss
+
+from src.models.vtr.ocr import OCRHead
+from src.utils.common import PositionalEncoding
+from src.models.embedders.vtr import VTREmbedder
+from src.utils.common import char2int
+
+
+def compute_ctc_loss(
```
Comment (borisshapa) [inline src/models/tasks.py:15]: I'd move this function to the `utils` directory. common.py or create new file losses.py

**Review by borisshapa — APPROVED** [review]

**Comment by borisshapa** [comment]
For the future: try to give more descriptive branch names than 'temp'.

**Review by SpirinEgor — APPROVED** [review]
